File: fastify_server/src/registry/ctx.ts
import type {
  HookCollector,
  HookFn,
  LoaderReport,
  ResolvedBootstrapConfig,
} from "../contract/index.js";
import { createLoaderReport } from "../contract/index.js";
import type { RuntimeAdapter } from "../adapters/adapter.js";

export interface RuntimeContext<Server = unknown> {
  registerInitHook(fn: HookFn<Server>): void;
  registerStartupHook(fn: HookFn<Server>): void;
  registerShutdownHook(fn: HookFn<Server>): void;
  getAdapter(): RuntimeAdapter<Server>;
  getConfig(): ResolvedBootstrapConfig;
  createReport(name: string): LoaderReport;
  logger: Logger;
}

export interface Logger {
  info(msg: string, ...meta: unknown[]): void;
  warn(msg: string, ...meta: unknown[]): void;
  error(msg: string, ...meta: unknown[]): void;
  debug(msg: string, ...meta: unknown[]): void;
}

export const consoleLogger: Logger = {
  info: (m, ...x) => console.log("[polyglot]", m, ...x),
  warn: (m, ...x) => console.warn("[polyglot]", m, ...x),
  error: (m, ...x) => console.error("[polyglot]", m, ...x),
  debug: (m, ...x) => {
    if (process.env.POLYGLOT_DEBUG) console.log("[polyglot:debug]", m, ...x);
  },
};

export function createContext<Server>(
  adapter: RuntimeAdapter<Server>,
  config: ResolvedBootstrapConfig,
  hooks: HookCollector<Server>,
  logger: Logger = consoleLogger,
): RuntimeContext<Server> {
  return {
    registerInitHook: (fn) => {
      hooks.init.push(fn);
    },
    registerStartupHook: (fn) => {
      hooks.startup.push(fn);
    },
    registerShutdownHook: (fn) => {
      hooks.shutdown.push(fn);
    },
    getAdapter: () => adapter,
    getConfig: () => config,
    createReport: createLoaderReport,
    logger,
  };
}

File: fastapi_server/fastapi_server/bootstrap/registry/ctx.py
from __future__ import annotations

import os
from dataclasses import dataclass
from typing import Any, Callable, Protocol

from fastapi_server.bootstrap.contract import (
    HookCollector,
    LoaderReport,
    ResolvedBootstrapConfig,
    create_loader_report,
)


class Logger(Protocol):
    def info(self, msg: str, *args: Any) -> None: ...
    def warn(self, msg: str, *args: Any) -> None: ...
    def error(self, msg: str, *args: Any) -> None: ...
    def debug(self, msg: str, *args: Any) -> None: ...


class StdoutLogger:
    def info(self, msg: str, *args: Any) -> None:
        print("[polyglot]", msg, *args)

    def warn(self, msg: str, *args: Any) -> None:
        print("[polyglot:warn]", msg, *args)

    def error(self, msg: str, *args: Any) -> None:
        print("[polyglot:error]", msg, *args)

    def debug(self, msg: str, *args: Any) -> None:
        if os.environ.get("POLYGLOT_DEBUG"):
            print("[polyglot:debug]", msg, *args)


HookFn = Callable[[Any, ResolvedBootstrapConfig], Any]


@dataclass
class RuntimeContext:
    adapter: Any
    config: ResolvedBootstrapConfig
    hooks: HookCollector
    logger: Logger

    def register_init_hook(self, fn: HookFn) -> None:
        self.hooks.init.append(fn)

    def register_startup_hook(self, fn: HookFn) -> None:
        self.hooks.startup.append(fn)

    def register_shutdown_hook(self, fn: HookFn) -> None:
        self.hooks.shutdown.append(fn)

    def get_adapter(self) -> Any:
        return self.adapter

    def get_config(self) -> ResolvedBootstrapConfig:
        return self.config

    def create_report(self, name: str) -> LoaderReport:
        return create_loader_report(name)


def create_context(
    adapter: Any,
    config: ResolvedBootstrapConfig,
    hooks: HookCollector,
    logger: Logger | None = None,
) -> RuntimeContext:
    return RuntimeContext(
        adapter=adapter,
        config=config,
        hooks=hooks,
        logger=logger or StdoutLogger(),
    )
